test/e2e/lua/dynamic_configuration.go

```diff
@@ -195,7 +193,11 @@ func enableDynamicConfiguration(kubeClientSet kubernetes.Interface) error {
 			args = append(args, "--enable-dynamic-configuration")
 			deployment.Spec.Template.Spec.Containers[0].Args = args
 			_, err := kubeClientSet.AppsV1beta1().Deployments("ingress-nginx").Update(deployment)
-			return err
+			if err != nil {
+				return err
+			}
+			time.Sleep(15 * time.Second)
```

Comment:
why does this necessary considering we wait for the pods to become "running" at https://github.com/kubernetes/ingress-nginx/blob/ee46f486c77665096f366eb215aa135032291b5a/test/e2e/framework/framework.go#L273 after this function executed?

Reply:
Because it matches on the old generation of pods and instantly returns

Reply:
Good catch, thanks!

--

I wonder if it would be better to use `WaitForNginxServer` here to assert for a dynamic feature specific Nginx configuration(i.e the presence of `upstream upstream_balancer`)? Similarly we can assert absence of `upstream upstream_balancer` in `disableDynamicConfiguration`. That would save us some seconds.

Reply:
Yeah, that's indeed a good idea.

I toyed a little around with checking if the deployments `Status.UpdatedReplicas` matches `Spec.Replicas` but that resulted in the instant return as well, apparently the controller sets `updatedReplicas` as soon as the new pod is in state creating and not when its ready.

I wonder what the best approach to check if a new generation of a deployment was fully rolled out is.

Reply:
Maybe as a more generic solution we could get the list of currently running pod names at https://github.com/kubernetes/ingress-nginx/blob/ee46f486c77665096f366eb215aa135032291b5a/test/e2e/framework/framework.go#L264 and then path them as a list to `WaitForPodsReady` and in that function on top of requiring given number of running pods also require that the names of pods are not in the list of old pods(if any given). That way before returning in `WaitForPodsReady` we would also guarantee that it is the new pods that are ready.